### PR TITLE
HttpPostRequestDecoder: retain instead of copy when first buf is last

### PR DESCRIPTION
**Pull request by fabienrenaud**
Motivations
-----------
There is no need to copy the "offered" ByteBuf in HttpPostRequestDecoder
when the first HttpContent ByteBuf is also the last (LastHttpContent) as
the full content can immediately be decoded. No extra bookeeping needed.

Modifications
-------------
In HttpPost(Standard|Multipart)RequestDecoder:
 - Retain the first ByteBuf when it is both the first HttpContent and last (LastHttpContent)
offered to the decoder.
 - Retain slices of the final buffers values
 - For non-multi-part content, URL decoding of Attributes' content ByteBuf no longer creates
a ByteBuf when there is nothing to decode. When decoding is necessary, a ByteBuf is created
using the allocator of the HttpContent's ByteBuf.

Results
-------
ByteBufs of FullHttpMessage decoded by HttpPostRequestDecoder are no longer
unnecessarily copied. Attributes are extracted as retained slices. When URL decoding
is necessary, the new ByteBuf is created using the allocator of the HttpContent's ByteBuf.

Partially addresses issue #10200

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java [hunk 1]
```diff
@@ -321,18 +321,19 @@ public InterfaceHttpData getBodyHttpData(String name) {
...
+        }
+
+        ByteBuf buf = content.content();
         // Maybe we should better not copy here for performance reasons but this will need
         // more care by the caller to release the content in a correct manner later
         // So maybe something to optimize on a later stage
-        ByteBuf buf = content.content();
         if (undecodedChunk == null) {
-            undecodedChunk = buf.copy();
+            undecodedChunk = isLastChunk ? buf.retain() : buf.copy();
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:333]: @fabienrenaud let's use `buf.retainedSlice()`

Comment (fabienrenaud) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java]: right...
done!

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java [hunk 2]
```diff
@@ -321,18 +321,19 @@ public InterfaceHttpData getBodyHttpData(String name) {
...
+        }
+
+        ByteBuf buf = content.content();
         // Maybe we should better not copy here for performance reasons but this will need
         // more care by the caller to release the content in a correct manner later
         // So maybe something to optimize on a later stage
-        ByteBuf buf = content.content();
         if (undecodedChunk == null) {
-            undecodedChunk = buf.copy();
+            undecodedChunk = isLastChunk ? buf.retainedSlice() : buf.copy();
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:333]: @fabienrenaud Can you add a comment that its only fine to not copy here as we will never call `undecodedChunk.writeBytes(...)` later on ?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java [hunk 3]
```diff
@@ -280,18 +280,19 @@ public InterfaceHttpData getBodyHttpData(String name) {
...
+        }
+
+        ByteBuf buf = content.content();
         // Maybe we should better not copy here for performance reasons but this will need
         // more care by the caller to release the content in a correct manner later
         // So maybe something to optimize on a later stage
-        ByteBuf buf = content.content();
         if (undecodedChunk == null) {
-            undecodedChunk = buf.copy();
+            undecodedChunk = isLastChunk ? buf.retainedSlice() : buf.copy();
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java:292]: same comment as above

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java (hunk 2 above)
Comment (fabienrenaud) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:333]: added

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java (hunk 3 above)
Comment (fabienrenaud) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java]: added

**Comment by fabienrenaud** [comment]
@normanmaurer Implemented a `ByteBuf decodeAttribute(ByteBuf b, Charset charset)` to avoid unnecessary copies when decoding non-multi-part attribute values.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@fabienrenaud please check... there is a memory leak. You can run with leak detection using `-Pleak`

**Comment by fabienrenaud** [comment]
@normanmaurer Leak is fixed. Was missing a `req.release()` in a multi-part test.

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java [hunk 4]
```diff
@@ -747,7 +769,7 @@ public void testNotLeakHeapBufferWhenWrapIllegalArgumentException() {
     }
 
     private static void testNotLeakWhenWrapIllegalArgumentException(ByteBuf buf) {
-        buf.writeCharSequence("==", CharsetUtil.US_ASCII);
+        buf.writeCharSequence("a=b&foo=%22bar%22&==", CharsetUtil.US_ASCII);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java:772]: @fabienrenaud why this change ?

Comment (fabienrenaud) [reply codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java:772]: the previous sequence was failing decoding immediately (on the first character). 
the new sequence causes the decoder to create a few new/sliced ByteBufs before failing the decoding of the last entry `==`.
in other words, the check now verifies that ByteBufs created before the exception are also properly released and do not leak.

Comment (normanmaurer) [reply codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java:772]: got it... thanks

**Comment by normanmaurer** [comment]
@fabienrenaud thanks a lot....

**Comment by fabienrenaud** [comment]
@normanmaurer This item should be in the list of "most important changes" in the release notes of 4.1.50 IMO. 
Upgrading to 4.1.50 may cause memory leaks in projects using this `HttpPostRequestDecoder` if they did/do not properly release the buffers returned by `HttpPostRequestDecoder`. 
The buffers returned prior to this change were always unpooled heap bufers so it is possible "projects out there" may have missed releasing buffers created by `HttpPostRequestDecoder`.

**Comment by doom369** [comment]
@fabienrenaud @normanmaurer true - https://github.com/netty/netty/issues/10281
